File: Nearest smaller neighbour/nearestSmallerNeighbour.cpp
#include <bits/stdc++.h>
using namespace std;

// Function to calculate nearest smaller neighbour for all array elements

void nearestSmallerNeighbour(int arr[], int N){
  
	stack<int> S;

	for (int i=0; i<N; i++){
    
		// remove the top of stack S till top element is greater than or equal to arr[i]
		
		while (!S.empty() && S.top() >= arr[i])
			S.pop();

		// condition for empty stack
		if (S.empty())
			cout <<-1<<" ";
			
		else //Else print the nearest smaller neighbour
			cout << S.top() << " ";

		S.push(arr[i]);
    
	}
}

int main(){
  
	int N = 5; // size of array
	
        int arr[] =  { 1, 2, 5, 3, 5 };
    
	nearestSmallerNeighbour(arr, N);
	
	return 0;
}

File: 16_Nearest smaller neighbour/nearestSmallerNeighbour.java
import java.io.*;
import java.util.*;
import java.lang.*;


class Main{
	//This function returns an array having the nearest smaller element for each number or -1 if it does not exists.
	public static int[] nearestSmallerNeighbour(int[] arr,int n){
		Stack<Integer> stack = new Stack<>();
		int[] ans = new int[n];
		for (int i=0;i<n;i++){
			while(!stack.isEmpty() && stack.peek() >= arr[i]) stack.pop();
			if (stack.isEmpty()) ans[i] = -1;
			else{
				ans[i] = stack.peek();
			}
			stack.push(arr[i]);
		}
		return ans;
	}
	//This function prints all the elements of an array on a single line.
	public static void printArray(int[] ans){
		for (int i=0;i<ans.length;i++){
			System.out.print(ans[i] + " ");
		}
		System.out.println();
	}
	//Main Function
	public static void main(String[] args){
		int n = 5;
		int[] arr = {1,2,5,3,5};
		int[] ans = nearestSmallerNeighbour(arr,n);
		printArray(ans);
	}
}